File: commands/stats/recentwins.py
# https://github.com/viktorexe/terristats-discord-bot
import discord
from discord.ext import commands
from discord import app_commands
from datetime import datetime, timezone, timedelta
from collections import Counter
from database import db

class RecentWins(commands.Cog):
    def __init__(self, bot):
        self.bot = bot

    @app_commands.command(name='recentwins', description='Show recent wins in last X hours')
    async def recentwins(self, interaction: discord.Interaction, hours: int):
        try:
            await interaction.response.defer()
            
            if hours <= 0 or hours > 168:  # Max 1 week
                embed = discord.Embed(
                    title="❌ Invalid Hours", 
                    description="Please enter hours between 1 and 168 (1 week).", 
                    color=0xff0000
                )
                await interaction.followup.send(embed=embed)
                return
            
            cutoff_time = datetime.now(timezone.utc) - timedelta(hours=hours)
            query = {'timestamp': {'$gte': cutoff_time}}
            
            pipeline = [
                {'$match': query},
                {'$facet': {
                    'stats': [{'$group': {
                        '_id': None,
                        'total': {'$sum': 1},
                        'contests': {'$sum': {'$cond': ['$contest', 1, 0]}},
                        'points': {'$sum': '$player_count'}
                    }}],
                    'clans': [{'$group': {'_id': '$clan_name', 'count': {'$sum': 1}}}, {'$sort': {'count': -1}}, {'$limit': 5}],
                    'maps': [{'$group': {'_id': '$map', 'count': {'$sum': 1}}}, {'$sort': {'count': -1}}, {'$limit': 3}],
                    'latest': [{'$sort': {'timestamp': -1}}, {'$limit': 5}, {'$project': {'clan_name': 1, 'map': 1, 'timestamp': 1}}]
                }}
            ]
            
            result = await db.db.clanwins.aggregate(pipeline).to_list(1)
            if not result or not result[0].get('stats'):
                embed = discord.Embed(
                    title=f"📅 Recent Wins ({hours}h)", 
                    description=f"No wins found in the last {hours} hours.", 
                    color=0xff0000
                )
                await interaction.followup.send(embed=embed)
                return
            
            data = result[0]
            stats = data['stats'][0] if data['stats'] else {}
            total_wins = stats.get('total', 0)
            contest_wins = stats.get('contests', 0)
            total_points = stats.get('points', 0)
            
            top_clans = [(c['_id'], c['count']) for c in data.get('clans', [])]
            top_maps = [(m['_id'], m['count']) for m in data.get('maps', [])]
            latest_wins = data.get('latest', [])
            
            embed = discord.Embed(
                title=f"Recent Wins ({hours}h)",
                description=f"Activity in the last {hours} hours",
                color=0x2b2d31
            )
            
            stats_text = f"Games: **{total_wins}** | Contest: **{contest_wins}**\n"
            stats_text += f"Points: **{total_points}** | Avg: **{total_wins/hours:.1f}**/hr"
            embed.add_field(name="Stats", value=stats_text, inline=False)
            
            if top_clans:
                clan_text = "\n".join([f"{i+1}. [{name}] - {count}" for i, (name, count) in enumerate(top_clans)])
                embed.add_field(name="Most Active Clans", value=clan_text, inline=True)
            
            if top_maps:
                map_text = "\n".join([f"{i+1}. {name} - {count}" for i, (name, count) in enumerate(top_maps)])
                embed.add_field(name="Popular Maps", value=map_text, inline=True)
            
            if latest_wins:
                latest_text = []
                for win in latest_wins:
                    win_time = win.get('timestamp', datetime.now(timezone.utc))
                    if win_time.tzinfo is None:
                        win_time = win_time.replace(tzinfo=timezone.utc)
                    time_ago = datetime.now(timezone.utc) - win_time
                    hours_ago = int(time_ago.total_seconds() / 3600)
                    mins_ago = int((time_ago.total_seconds() % 3600) / 60)
                    
                    if hours_ago > 0:
                        time_str = f"{hours_ago}h"
                    else:
                        time_str = f"{mins_ago}m"
                    
                    latest_text.append(f"[{win.get('clan_name', 'Unknown')}] {win.get('map', 'Unknown')} - {time_str}")
                
                embed.add_field(name="Latest", value="\n".join(latest_text), inline=False)
            
            total_games = await db.db.clanwins.count_documents({})
            embed.set_footer(text=f"Data from 18 Nov 2025 | {total_games:,} games tracked")
            await interaction.followup.send(embed=embed)
            
        except Exception as e:
            print(f"💥 RECENTWINS ERROR: {e}")
            try:
                await interaction.followup.send(embed=discord.Embed(
                    title="Error", description="Failed to fetch recent wins.", color=0xff0000
                ))
            except:
                pass

async def setup(bot):
    await bot.add_cog(RecentWins(bot))

File: commands/stats/recentwins.js
// https://github.com/viktorexe/terristats-discord-bot
// /recentwins - all activity across every clan in the last X hours (max 168).
const { SlashCommandBuilder, EmbedBuilder } = require('discord.js');
const db = require('../../database');
const { gamesFooter } = require('../../utils');

module.exports = {
  data: new SlashCommandBuilder()
    .setName('recentwins')
    .setDescription('Show recent wins in last X hours')
    .addIntegerOption((o) =>
      o.setName('hours').setDescription('Look-back window (1-168)').setRequired(true),
    ),

  async execute(interaction) {
    try {
      await interaction.deferReply();

      const hours = interaction.options.getInteger('hours');

      if (hours <= 0 || hours > 168) {
        await interaction.editReply({
          embeds: [
            new EmbedBuilder()
              .setTitle('❌ Invalid Hours')
              .setDescription('Please enter hours between 1 and 168 (1 week).')
              .setColor(0xff0000),
          ],
        });
        return;
      }

      const cutoff = new Date(Date.now() - hours * 3600 * 1000);
      const query = { timestamp: { $gte: cutoff } };

      const pipeline = [
        { $match: query },
        {
          $facet: {
            stats: [
              {
                $group: {
                  _id: null,
                  total: { $sum: 1 },
                  contests: { $sum: { $cond: ['$contest', 1, 0] } },
                  points: { $sum: '$player_count' },
                },
              },
            ],
            clans: [
              { $group: { _id: '$clan_name', count: { $sum: 1 } } },
              { $sort: { count: -1 } },
              { $limit: 5 },
            ],
            maps: [
              { $group: { _id: '$map', count: { $sum: 1 } } },
              { $sort: { count: -1 } },
              { $limit: 3 },
            ],
            latest: [
              { $sort: { timestamp: -1 } },
              { $limit: 5 },
              { $project: { clan_name: 1, map: 1, timestamp: 1 } },
            ],
          },
        },
      ];

      const result = await db.collection('clanwins').aggregate(pipeline).toArray();
      if (!result.length || !result[0].stats.length) {
        await interaction.editReply({
          embeds: [
            new EmbedBuilder()
              .setTitle(`📅 Recent Wins (${hours}h)`)
              .setDescription(`No wins found in the last ${hours} hours.`)
              .setColor(0xff0000),
          ],
        });
        return;
      }

      const data = result[0];
      const stats = data.stats[0] || {};
      const totalWins = stats.total || 0;
      const contestWins = stats.contests || 0;
      const totalPoints = stats.points || 0;

      const topClans = data.clans.map((c) => [c._id, c.count]);
      const topMaps = data.maps.map((m) => [m._id, m.count]);
      const latestWins = data.latest || [];

      const embed = new EmbedBuilder()
        .setTitle(`Recent Wins (${hours}h)`)
        .setDescription(`Activity in the last ${hours} hours`)
        .setColor(0x2b2d31);

      let statsText = `Games: **${totalWins}** | Contest: **${contestWins}**\n`;
      statsText += `Points: **${totalPoints}** | Avg: **${(totalWins / hours).toFixed(1)}**/hr`;
      embed.addFields({ name: 'Stats', value: statsText, inline: false });

      if (topClans.length) {
        embed.addFields({
          name: 'Most Active Clans',
          value: topClans.map(([n, c], i) => `${i + 1}. [${n}] - ${c}`).join('\n'),
          inline: true,
        });
      }
      if (topMaps.length) {
        embed.addFields({
          name: 'Popular Maps',
          value: topMaps.map(([n, c], i) => `${i + 1}. ${n} - ${c}`).join('\n'),
          inline: true,
        });
      }

      if (latestWins.length) {
        const latestText = latestWins.map((win) => {
          const winTime = new Date(win.timestamp || Date.now());
          const diffMs = Date.now() - winTime.getTime();
          const hoursAgo = Math.floor(diffMs / 3600000);
          const minsAgo = Math.floor((diffMs % 3600000) / 60000);
          const timeStr = hoursAgo > 0 ? `${hoursAgo}h` : `${minsAgo}m`;
          return `[${win.clan_name || 'Unknown'}] ${win.map || 'Unknown'} - ${timeStr}`;
        });
        embed.addFields({ name: 'Latest', value: latestText.join('\n'), inline: false });
      }

      const totalGames = await db.collection('clanwins').countDocuments({});
      embed.setFooter({ text: gamesFooter(totalGames) });
      await interaction.editReply({ embeds: [embed] });
    } catch (err) {
      console.log(`RECENTWINS ERROR: ${err.message}`);
      try {
        await interaction.editReply({
          embeds: [
            new EmbedBuilder()
              .setTitle('Error')
              .setDescription('Failed to fetch recent wins.')
              .setColor(0xff0000),
          ],
        });
      } catch (e) {
        // ignore
      }
    }
  },
};
